language/diem-vm/src/diem_transaction_executor.rs

```diff
@@ -590,6 +591,29 @@ impl DiemVM {
         ))
     }
 
+    fn preload_cache(
+        signature_verified_block: &[PreprocessedTransaction],
+        data_cache: &mut StateViewCache,
+    ) {
+        // generate a collection of addresses
+        let mut addresses_to_preload = Vec::new();
+        for txn in signature_verified_block {
+            if let PreprocessedTransaction::UserTransaction(txn) = txn {
+                if let TransactionPayload::Script(script) = txn.payload() {
+                    addresses_to_preload.push(txn.sender());
+
+                    for arg in script.args() {
+                        if let TransactionArgument::Address(address) = arg {
+                            addresses_to_preload.push(*address);
+                        }
+                    }
+                }
+            }
+        }
+
+        addresses_to_preload.into_par_iter().map(|addr| data_cache.get(&AccessPath::new(addr, Vec::new())));
```

Comment:
Do you mean, try to limit the number of threads that rayon uses here, or experiment with different numbers?

Reply:
I'm not sure what @runtian-zhou meant, but in my opinion, we may only want one thread for prefetching (in parallel with a thread for execution) anyway. We could do some benchmarking to check that, but my intuition is that until we have parallel execution, we want to prioritize the prefetching in the order of the transactions in the block, since they'll be executed in that order. If we fork off a bunch of threads at once, the extra work may slow down the time to get the data for the first transaction, etc.

Reply:
Sry lemme be clear on the two alternatives:
1. Spawn a detached thread to prefetch the account data while executing txns (non-blocking)
2. Spawn a number of threads to prefetch all account data before executing txns (blocking)

This implementation is looking for the second alternative. What I was trying to ask is the number of threads being used over there. Since this is a blocking operation, we need to spawn the right number of threads to maximize the perf over there. That number is something we might want to tune in this PR or in the PR to come.

With that being said, if we chose to go for the first alternative, the granularity of threads will no longer be an issue. On the other hand, there will be issues around caching, as you can no longer simply write to `StateViewCache`. My suggestion is to get the data from `dyn StateView` instead cuz the `VerifiedStateView` that the `DiemVM` uses actually implements a cache in the backend.

Reply:
Thanks, @runtian-zhou -- that helps a lot. My intuition is that option 1 will be faster, but I've learned that my intuition about performance issues in unreliable. It seems to me like we should try both and see which one is faster, including whatever tuning helps option 2 work best (assuming there is a way to make option 1 work).